### PR TITLE
Feature Flag Scheduling

### PR DESCRIPTION
**Pull request by mknowlton89**
### Features and Changes
This PR will allow organizations to build schedules for enabling/disabling feature flag rules. Currently, the UI only supports setting a start date and end date, and the end date must be greater than the start date. However, the feature was built in a way where in the future, we can allow organizations to build schedules with more than two schedule rules.

To do this, a user can go to a feature flag and create a rule. Below the condition targeting, a user can create schedule rules. Behind the scenes, these rules are used by `getFeatureDefinitions` and if a scheduleRule has not been met, that feature is filtered out.

Additionally, this PR includes work for a cron job that runs every minute, and gets features that have a "nextScheduledUpdate" property that is in the past. This will fire a webhook to invalidate an organizations cache, to ensure a feature flag that is scheduled to launch/end at a specific date and time does so correctly, even when the status of a feature/rule is cached.

### Dependencies
 N/A

### Testing

- [x] Build an override rule where the userId is equal to your current userId, and then add schedule rules for the rule to launch immediately, and end in 2 weeks. Ensure that when you save the rule and go to the next page, the feature is evaluated to "on".
- [x] Build an override rule where the userId is equal to your current userId, and then add schedule rules for the rule to launch in a day, and end in 2 weeks. Ensure that when you save the rule and go to the next page, the feature is evaluated to "off" and the rule is shown as "disabled" with a note that the feature will be enabled in 1 day.
- [x] Build an override rule where the userId is equal to your current userId, and then add schedule rules for the rule to launch 1 day ago, and end in 2 weeks. Ensure that when you save the rule and go to the next page, the feature is evaluated to "on" and the rule has a note that says the feature will be disabled in two weeks.
- [x] Build an override rule where the userId is equal to your current userId, and then add schedule rules for the rule to launch 1 day ago, and manually. Ensure that when you save the rule and go to the next page, the feature is evaluated to "on" and there is no note about when the feature will be disabled.
- [x] Build an override rule where the userId is equal to your current userId, and then add schedule rules for the rule to launch 1 week ago, and end in 1 day ago. Ensure that when you save the rule and go to the next page, the feature is evaluated to "off" and the rule has a note that the feature was disabled by a schedule rule yesterday.
- [x] Build an override rule where the userId is equal to your current userId, and then add schedule rules for the rule to launch 5 muntes, and end in 2 weeks. Ensure that when you save the rule and go to the next page, the feature is evaluated to "off" and the rule has a note saying the feature will be enabled in 5 minutes. Then, wait 5 minutes, refresh the page and ensure the feature is now evaluated to "on".

#### Note: There is a capitalization error in the screenshots below (Launch rule vs Disable Rule) - this has been addressed.

Screen Shot 2022-12-05 at 9 16 16 AM

Screen Shot 2022-12-05 at 9 16 23 AM

Screen Shot 2022-12-05 at 9 16 30 AM

Screen Shot 2022-12-05 at 9 16 50 AM

Screen Shot 2022-12-05 at 9 16 59 AM

Screen Shot 2022-12-05 at 10 16 08 AM

Screen Shot 2022-12-05 at 9 18 30 AM

Screen Shot 2022-12-05 at 10 18 24 AM

### CODE REVIEW

File: packages/back-end/src/services/features.ts [hunk 1]
```diff
@@ -393,3 +419,37 @@ export function getApiFeatureObj(
...
+  }
+
+  const dates: string[] = [];
+
+  if (envSettings) {
+    for (const env in envSettings) {
+      const rules = envSettings[env].rules;
+
+      rules.forEach((rule: FeatureRule) => {
+        if (rule.scheduleRules) {
```
Comment (mknowlton89) [inline packages/back-end/src/services/features.ts:437]: I think I can simplify this by spreading in the scheduleRules array into the dates array, rather than just looping through it.

File: packages/front-end/components/Features/Rule.tsx [hunk 2]
```diff
@@ -60,6 +64,41 @@ export const Rule = forwardRef<HTMLDivElement, RuleProps>(
       permissions.check("manageFeatures", feature.project) &&
       permissions.check("createFeatureDrafts", feature.project);
 
+    const hasConditions = rule.condition && rule.condition !== "{}";
+
+    let upcomingScheduleRule: ScheduleRule | null = null;
```
Comment (mknowlton89) [inline packages/front-end/components/Features/Rule.tsx:69]: This is done in two places in the app, I should explore making this a service function.

File: packages/front-end/components/Features/Rule.tsx [hunk 3]
```diff
@@ -222,7 +300,7 @@ export const Rule = forwardRef<HTMLDivElement, RuleProps>(
                 }}
               ></div>
             )}
-            {rule.condition && rule.condition !== "{}" && (
+            {hasConditions && (
```
Comment (mknowlton89) [inline packages/front-end/components/Features/Rule.tsx:303]: Creating this condition is no longer needed as this logic is now only used in this one spot.

File: packages/front-end/components/Features/RuleModal.tsx [hunk 4]
```diff
@@ -68,6 +73,16 @@ export default function RuleModal({
       header={rules[i] ? "Edit Override Rule" : "New Override Rule"}
       submit={form.handleSubmit(async (values) => {
         const ruleAction = i === rules.length ? "add" : "edit";
+
+        if (values.scheduleRules.length) {
```
Comment (mknowlton89) [inline packages/front-end/components/Features/RuleModal.tsx:77]: Is this the right structure for the timestamp field?

File: packages/front-end/components/Features/ScheduleInputs.tsx [hunk 5]
```diff
@@ -0,0 +1,142 @@
+import { ScheduleRule } from "back-end/types/feature";
+import React, { useEffect, useState } from "react";
+import Field from "../Forms/Field";
+import SelectField from "../Forms/SelectField";
+import { GBAddCircle } from "../Icons";
+
+interface Props {
+  defaultValue: ScheduleRule[];
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (mknowlton89) [inline packages/front-end/components/Features/ScheduleInputs.tsx:9]: Need to update this any to the actual shape.

File: packages/front-end/components/Features/ScheduleInputs.tsx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export default function ScheduleInputs(props: Props) {
+  const [rules, setRules] = useState(props.defaultValue);
+
+  useEffect(() => {
+    props.onChange(rules);
+  }, [props, props.defaultValue, rules]);
+
+  const options = [
+    { label: "enabled", value: "true" },
```
Comment (mknowlton89) [inline packages/front-end/components/Features/ScheduleInputs.tsx:35]: Having to set the value here to a string is a bit weird. Is there a better way to do this?

File: packages/front-end/components/Features/Rule.tsx [hunk 7]
```diff
@@ -60,6 +64,41 @@ export const Rule = forwardRef<HTMLDivElement, RuleProps>(
...
+
+    const nextRuleIndex = sortedScheduleRules.findIndex(
+      (rule) => currentDate < new Date(rule.timestamp).valueOf()
+    );
+
+    if (nextRuleIndex > -1) {
+      upcomingScheduleRule = sortedScheduleRules[nextRuleIndex];
+    }
+
+    let currentScheduleStatus;
```
Comment (mknowlton89) [inline packages/front-end/components/Features/Rule.tsx:86]: Should I set this as a state variable instead?

File: packages/front-end/components/Features/Rule.tsx [hunk 8]
```diff
@@ -60,6 +64,41 @@ export const Rule = forwardRef<HTMLDivElement, RuleProps>(
...
+      currentScheduleStatus = rule.scheduleRules[rule.scheduleRules.length - 1]
+        .enableFeature
+        ? "enabled"
+        : "disabled";
+    }
+    if (!rule.scheduleRules || !rule.scheduleRules.length) {
+      currentScheduleStatus = "enabled";
+    }
+
+    const ruleDisabled =
```
Comment (mknowlton89) [inline packages/front-end/components/Features/Rule.tsx:98]: Can I update this to also include if `!rule.enabled`?

Comment (mknowlton89) [reply packages/front-end/components/Features/Rule.tsx:98]: No, it breaks in the event that a rule has been actually "Disabled" by the user AND the current scheduling rule has the feature "disabled". Ultimately, I just need to differentiate between enable/disable in the manual sense, and the "status" that is derived from the schedule rules.

File: packages/front-end/components/Features/RuleModal.tsx [hunk 9]
```diff
@@ -68,6 +73,24 @@ export default function RuleModal({
...
```
Comment (mknowlton89) [inline packages/front-end/components/Features/RuleModal.tsx:77]: This will need to be updated if we want to support more than a start date and end date.

File: packages/front-end/components/Features/RuleStatusPill.tsx [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  upcomingScheduleRule: ScheduleRule | null;
+  currentScheduleStatus: "enabled" | "disabled";
+};
+
+export default function RuleStatusPill({
+  rule,
+  upcomingScheduleRule,
+  currentScheduleStatus,
+}: Props) {
+  if (!rule.enabled) {
```
Comment (mknowlton89) [inline packages/front-end/components/Features/RuleStatusPill.tsx:15]: Rather than having all of these blocks, can I create a variable `pillText` and build a switch/case statement and then just feed that into the return?

File: packages/front-end/services/scheduleRules.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  if (nextRuleIndex > -1) {
+    return rule.scheduleRules[nextRuleIndex];
+  }
+}
+
+export function getCurrentStatus(
+  rule: FeatureRule,
+  upcomingScheduleRule: ScheduleRule | null
+): "enabled" | "disabled" {
+  console.log("upcomingScheduleRule", upcomingScheduleRule);
```
Comment (mknowlton89) [inline packages/front-end/services/scheduleRules.ts:33]: Remove this

File: packages/back-end/src/services/features.ts [hunk 12]
```diff
@@ -81,6 +82,33 @@ export function getFeatureDefinition({
     rules:
       rules
         ?.filter((r) => r.enabled)
+        ?.filter((r) => {
+          //Filter our rules that have schedulingRules if the current scheduleRule sets the feature to disabled.
+          if (!r.scheduleRules || !r.scheduleRules.length) {
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/services/features.ts:87]: nit: can be simplified to `!r?.scheduleRules?.length`

File: packages/back-end/src/services/features.ts [hunk 13]
```diff
@@ -81,6 +82,33 @@ export function getFeatureDefinition({
...
+          // Loop through the list of rules, and find the next rule to be applied
+          const nextRuleIndex = r.scheduleRules.findIndex(
+            (rule) =>
+              rule.timestamp !== null &&
+              new Date(rule.timestamp).valueOf() > currentDate
+          );
+
+          if (nextRuleIndex === -1) {
+            if (
+              r.scheduleRules[r.scheduleRules.length - 1].timestamp === null
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/services/features.ts:102]: nit, for fun: I think we can now get last element using `myarray.at(-1)`

File: packages/back-end/src/services/features.ts [hunk 14]
```diff
@@ -397,6 +425,40 @@ export function getApiFeatureObj(
...
+export function getNextScheduledUpdate(
+  envSettings: Record<string, FeatureEnvironment>
+): Date | null {
+  if (!envSettings) {
+    return null;
+  }
+
+  const dates: string[] = [];
+
+  if (envSettings) {
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/services/features.ts:437]: this check is not needed since we return on line 432

File: packages/back-end/src/services/features.ts (hunk 13 above)
Comment (mknowlton89) [reply packages/back-end/src/services/features.ts:102]: Today I learned!

File: packages/back-end/src/models/FeatureModel.ts [hunk 15]
```diff
@@ -48,6 +53,12 @@ const featureSchema = new mongoose.Schema({
         },
       ],
       namespace: {},
+      scheduleRules: [
+        {
+          timestamp: String,
```
[the remark this answers is not in the document]
Comment (mknowlton89) [reply packages/back-end/src/models/FeatureModel.ts:58]: I've gone back and forth on this, but I'm not sure I have enough context to know which is the right direction.

Initially I thought about adding it so we can change the shape of the `nextScheduledUpdate` property from a single date, to a `ScheduleRule` object. The idea being it could make getting data about the next scheduled update a bit easier. Currently, we have to loop through the array of rules and find the first rule where the timestamp is greater than the currentDate.

If we added the _id (or, just an `id` property, we could instead get the scheduleRule.id from the `nextScheduledUpdate` and then access that index in the scheduleRules array. 

I left it out given that it could increase complexity of the cron job's query. But open to discussion on if other's think that is a good enough reason to not add the `_id` or just an `id` property.

[the remark this answers is not in the document]
Comment (bryce-fitzsimons) [reply packages/back-end/src/models/FeatureModel.ts:58]: Sure, if we are targeting these ScheduleRules by id somewhere (cron, etc), then it does make sense to keep the field I think.

**Comment by bryce-fitzsimons** [comment]
Perhaps not specific to your PR, but might be good to address: When editing an override rule, I can't change the type of rule. Perhaps this is intentional, but the UI suggests that the field is actually editable:
image


There are some one-off rules in `global.scss` that add graying out for \<select\> fields. Perhaps you could piggyback on these rules, extend them to your input field?

```
.gb-select-wrapper.disabled {
  cursor: not-allowed;
}
.gb-select--is-disabled {
  .gb-select__control {
    background-color: var(--disabled-background) !important;
  }
  .gb-select__single-value {
    color: var(--text-color-muted) !important;
  }
}
```

**Comment by mknowlton89** [comment]
> Perhaps not specific to your PR, but might be good to address: When editing an override rule, I can't change the type of rule. Perhaps this is intentional, but the UI suggests that the field is actually editable: image
> 
> There are some one-off rules in `global.scss` that add graying out for <select> fields. Perhaps you could piggyback on these rules, extend them to your input field?
> 
> ```
> .gb-select-wrapper.disabled {
>   cursor: not-allowed;
> }
> .gb-select--is-disabled {
>   .gb-select__control {
>     background-color: var(--disabled-background) !important;
>   }
>   .gb-select__single-value {
>     color: var(--text-color-muted) !important;
>   }
> }
> ```

I just created this issue for this - https://github.com/growthbook/growthbook/issues/822

File: packages/back-end/src/jobs/updateScheduledFeatures.ts [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+    featureId,
+  });
+  job.schedule(new Date());
+  await job.save();
+}
+
+export default async function (agenda: Agenda) {
+  agenda.define(QUEUE_FEATURE_UPDATES, async () => {
+    const featureIds = (
+      await FeatureModel.find({
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/jobs/updateScheduledFeatures.ts:40]: Not a big deal, but it does feel like .find() should be happening from within FeatureModel.ts

File: packages/back-end/src/jobs/updateScheduledFeatures.ts [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+  // Get the feature from the DB
+  const featureId = job.attrs.data?.featureId;
+  if (!featureId) return;
+
+  const log = logger.child({
+    cron: "updateSingleFeature",
+    featureId,
+  });
+
+  const feature = await FeatureModel.findOne({
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/jobs/updateScheduledFeatures.ts:72]: same here

File: packages/back-end/src/jobs/updateScheduledFeatures.ts [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+      feature.project || ""
+    );
+
+    // Then, we'll need to recalculate the feature's new nextScheduledUpdate and set it
+    const nextScheduledUpdate = getNextScheduledUpdate(
+      feature.environmentSettings || {}
+    );
+
+    // And finally, we'll need to update the feature with the new nextScheduledUpdate
+    await FeatureModel.updateOne(
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/jobs/updateScheduledFeatures.ts:91]: same

File: packages/back-end/src/models/FeatureModel.ts [hunk 19]
```diff
@@ -59,7 +70,10 @@ featureSchema.index({ id: 1, organization: 1 }, { unique: true });
 
 type FeatureDocument = mongoose.Document & LegacyFeatureInterface;
 
-const FeatureModel = mongoose.model<FeatureDocument>("Feature", featureSchema);
+export const FeatureModel = mongoose.model<FeatureDocument>(
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/models/FeatureModel.ts:73]: same here. Probably best to keep the mongoose model private and rely on public methods to access it.

File: packages/front-end/components/Features/Rule.tsx [hunk 20]
```diff
@@ -60,6 +62,18 @@ export const Rule = forwardRef<HTMLDivElement, RuleProps>(
...
+    const upcomingScheduleRule = getUpcomingScheduleRule(rule);
+
+    const scheduleCompletedAndDisabled =
+      !upcomingScheduleRule &&
+      rule?.scheduleRules?.length &&
+      rule.scheduleRules.at(-1)?.timestamp !== null;
+
+    const ruleDisabled =
+      scheduleCompletedAndDisabled ||
+      upcomingScheduleRule?.enableFeature ||
```
Comment (bryce-fitzsimons) [inline packages/front-end/components/Features/Rule.tsx:74]: Question for my own knowledge. Is it possible for a rule to be currently enabled, but also have an upcoming "enableFeature" schedule?

**Review by bryce-fitzsimons — DISMISSED** [review]
Conditionally approved. Overall looks good, just had a few questions/comments. If you would like to address them please feel free, otherwise feel free to disregard if you've already cleared those topics.

File: packages/back-end/src/jobs/updateScheduledFeatures.ts (hunk 17 above)
Comment (jdorn) [inline packages/back-end/src/jobs/updateScheduledFeatures.ts:72]: We always need an organization id when looking up features.  Ids alone are not unique

File: packages/back-end/src/jobs/updateScheduledFeatures.ts [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+      {
+        id: featureId,
+      },
```
Comment (jdorn) [inline packages/back-end/src/jobs/updateScheduledFeatures.ts:94]: Same here, we need an organization id here.

File: packages/back-end/src/jobs/updateScheduledFeatures.ts [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+  await updateFeatureJob.save();
+}
+
+async function queueFeatureUpdate(agenda: Agenda, featureId: string) {
+  const job = agenda.create(UPDATE_SINGLE_FEATURE, {
+    featureId,
+  }) as UpdateSingleFeatureJob;
+
+  job.unique({
+    featureId,
```
Comment (jdorn) [inline packages/back-end/src/jobs/updateScheduledFeatures.ts:31]: Feature id alone is not unique.  Also need organization id

File: packages/back-end/src/services/features.ts [hunk 23]
```diff
@@ -81,6 +82,30 @@ export function getFeatureDefinition({
...
+
+          if (nextRuleIndex === -1) {
+            if (r.scheduleRules.at(-1)?.timestamp === null) {
+              return true;
+            } else {
+              return r.scheduleRules.at(-1)?.enableFeature;
+            }
+          }
+
+          return !r.scheduleRules[nextRuleIndex].enableFeature;
```
Comment (jdorn) [inline packages/back-end/src/services/features.ts:107]: This logic doesn't feel quite right.  What if the scheduledRules array is not sorted?  What if there is only 1 rule in the future that says "enable"?

This logic should live in it's own util function and have unit tests that cover all of the edge cases.  Even if the code above does work and covers all the edge cases, there's no way to know just by looking at it.

```ts
getCurrentEnabledState(r.scheduledRules, currentDate)
```

**Review by jdorn — CHANGES_REQUESTED** [review]

File: packages/front-end/components/Features/Rule.tsx (hunk 20 above)
Comment (mknowlton89) [reply packages/front-end/components/Features/Rule.tsx:74]: @bryce-fitzsimons Yep! The `enableFeature` property is a boolean, so it's possible that a rule is enabled, and the upcoming rule's `enableFeature` value is `false`. This would indicate the rule is scheduled to be disabled at a date in the future.

File: packages/back-end/src/util/scheduleRules.ts [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+  scheduleRules: ScheduleRule[],
+  currentDate: Date
+) => {
+  // If a feature doesn't have any rules, don't filter it out
+  if (!scheduleRules.length) {
+    return true;
+  }
+
+  // Ensure the scheduleRules array is always sorted earliest to latest
+  const sortedRules = scheduleRules.sort((a, b) => {
```
Comment (mknowlton89) [inline packages/back-end/src/util/scheduleRules.ts:13]: @jdorn @bryce-fitzsimons I moved this to a util function and also added some logic to sort the `scheduleRules` by date, which requires some colorful logic.

Some background:
- Currently, a `scheduleRules` array can only have a length of 2, or be null, we don't support an array with a length of 1. 
- We do have some protections on the front end so the end date can't be before the start date, but I'm sure there are gotchas there, so I like the idea of sorting on the backend.
- This was built with a future in mind that would allow more than just a start/end date. For example, a user could start a feature immediately, disable it on 12/30/22, and then re-enable it on 1/05/23. The current logic does support that.

I've outlined what potential scheduleRule arrays could look like below.
```
// This is a rule that will launch immediately after being published, and turn off on 12/30/22.
[
{enableFeature: true, timestamp: null},
{enableFeature: false, timestamp: "2022-12-30T13:00:00.000Z"},
]

// This is a rule that will turn on, on 12/01/22 and will only turn off when the org manually disables it.
[
{enableFeature: true, timestamp: "2022-12-30T13:00:00.000Z"},
{enableFeature: false, timestamp: null },
]

// This is what a scheduleRule array looks like if the org wants it to start on 12/01/22 and turn off on 12/30/22

[
{enableFeature: true, timestamp: "2022-12-01T13:00:00.000Z"},
{enableFeature: false, timestamp: "2022-12-30T13:00:00.000Z" },
]
```

The colorful logic comes when we attempt to sort the array by date, when we do the date compare, we can't compare the date to `null`. So currently, the sort logic says, if a date is null, and the null date's corresponding `enableFeature` is true, that means this is a start rule, so I picked a really early date. But, if the date is null, and the date's corresponding `enableFeature` is false, that means it's an end date, so I picked a date in the far future.

While this works, it feels hacky. Have you come across a pattern you like to handle this sort of thing before?

File: packages/back-end/src/util/scheduleRules.ts [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+      dateA = new Date(a.timestamp);
+    }
+
+    if (b.timestamp === null) {
+      dateB = b.enableFeature ? new Date("01-01-1971") : new Date("01-01-2090"); //TODO: Find a better way to handle this
+    } else {
+      dateB = new Date(b.timestamp);
+    }
+
+    return dateA.valueOf() - dateB.valueOf();
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/util/scheduleRules.ts:28]: nit: don't need valueOf(), just dateA - dateB

File: packages/back-end/src/util/scheduleRules.ts [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+    } else {
+      dateA = new Date(a.timestamp);
+    }
+
+    if (b.timestamp === null) {
+      dateB = b.enableFeature ? new Date("01-01-1971") : new Date("01-01-2090"); //TODO: Find a better way to handle this
+    } else {
+      dateB = new Date(b.timestamp);
+    }
+
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/util/scheduleRules.ts:27]: This does feel confusing. I'm having some trouble thinking of what's supposed to happen in all the different scenarios. That said, it might help to do something like this (I'm sure my logic is completely incorrect, but the idea...):
```
if (!a.timestamp && !b.timestamp) {
  return 0;
}
if (!a.timestamp) {
  return a.enableFeature ? 1 : -1;
}
if (!b.timestamp) {
 return b.enableFeature ? -1 : 1;
}
return (new Date(a.timestamp)) - (new Date(b.timestamp));
```

File: packages/front-end/components/Features/RuleModal.tsx [hunk 27]
```diff
@@ -69,6 +74,26 @@ export default function RuleModal({
...
```
Comment (bryce-fitzsimons) [inline packages/front-end/components/Features/RuleModal.tsx:78]: I think you can remove the length check and just do:
`values.scheduleRules?.foreach(`

**Review by bryce-fitzsimons — APPROVED** [review]
conditionally approved. I had some inline feedback, but I think it's pretty good as-is, if you'd like to launch.

File: packages/back-end/src/util/scheduleRules.ts (hunk 26 above)
Comment (mknowlton89) [reply packages/back-end/src/util/scheduleRules.ts:27]: I just threw this in and tested it and it worked perfectly. I'm still wrapping my head around it, and I'm going to push on this logic with some crazy use cases. Thanks for the suggestion!

File: packages/front-end/components/Features/ScheduleInputs.tsx [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+
+const getLocalDateTime = (rawDateTime: string) => {
+  if (!rawDateTime) {
+    return "";
+  }
+  const utcDateTime = new Date(rawDateTime);
+
+  // We need to adjust for timezone/daylight savings time before converting to ISO String to pass into datetime-local field
+  utcDateTime.setHours(
+    utcDateTime.getHours() - new Date(rawDateTime).getTimezoneOffset() / 60
```
Comment (jdorn) [inline packages/front-end/components/Features/ScheduleInputs.tsx:21]: Timezone offset is not always a multiple of 60. Some timezones are offset by 15 or 30 minutes, so set hours might not work properly. Timezones are really hard and there are so many edge cases. I highly recommend using a library instead of doing this yourself. We have the date-fns library loaded already if that helps.

File: packages/front-end/components/Features/ScheduleInputs.tsx [hunk 29]
```diff
@@ -0,0 +1,180 @@
...
+                  <div className="col-sm-12 col-md mb-2 pl-2 pr-2">
+                    <SelectField
+                      name="date-operator"
+                      value={
+                        timestamp !== null
+                          ? "at a specific date and time"
+                          : enableFeature
+                          ? "immediately"
+                          : "manually"
+                      }
```
Comment (jdorn) [inline packages/front-end/components/Features/ScheduleInputs.tsx:111]: This is not a good way to determine the value of the select.

File: packages/front-end/components/Features/ScheduleInputs.tsx [hunk 30]
```diff
@@ -0,0 +1,180 @@
...
+                      options={(i + 1) % 2 === 1 ? launchOptions : endOptions}
```
Comment (jdorn) [inline packages/front-end/components/Features/ScheduleInputs.tsx:112]: Same here. Not a good way to figure out which options to show. Having this in a ul with a loop is causing the code to be very confusing. Since there are always exactly 2 inputs, it's probably better to get rid of the loop entirely and just hardcore both here. Will get rid of all the ternary and conditional logic.

File: packages/front-end/services/scheduleRules.ts [hunk 31]
```diff
@@ -0,0 +1,24 @@
...
+export function getUpcomingScheduleRule(
+  rule: FeatureRule
+): ScheduleRule | null {
+  if (!rule.scheduleRules || !rule.scheduleRules.length) {
+    return null;
+  }
+  const currentDate = new Date().valueOf();
+
+  const nextRuleIndex = rule.scheduleRules.findIndex(
+    (rule) =>
```
Comment (jdorn) [inline packages/front-end/services/scheduleRules.ts:12]: Calling this `rule` is a little confusing since you already have a rule variable that means something else.

Comment (mknowlton89) [reply packages/front-end/services/scheduleRules.ts:12]: Updated to `scheduleRule` to be more specific.

File: packages/back-end/types/feature.d.ts [hunk 32]
```diff
@@ -46,12 +47,16 @@ export interface FeatureInterface {
     publishedBy: UserRef;
   };
 }
-
+type ScheduleRule = {
+  timestamp: string | null;
+  enableFeature: boolean;
```
[the remark this answers is not in the document]
Comment (mknowlton89) [reply packages/back-end/types/feature.d.ts:52]: I'm leaning towards maybe `ruleStatus`.

```
type ScheduleRule = {
   timestamp: string | null;
   ruleStatus: "on" | "off";
```

File: packages/front-end/components/Features/ScheduleInputs.tsx (hunk 28 above)
Comment (mknowlton89) [reply packages/front-end/components/Features/ScheduleInputs.tsx]: Thank you for suggesting `date-fns`. I've not used that previously but that was exactly what was needed in order to format the date in a way that the `datetime-local` input needed.

File: packages/back-end/src/util/scheduleRules.ts [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+export const getCurrentEnabledState = (
+  scheduledRules: ScheduleRule[],
+  now: Date
+) => {
+  const sorted = scheduledRules
+    // Remove nulls
+    .filter((s) => s.timestamp)
+    // Convert timestamps to date objects
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    .map((s: any) => ({
```
Comment (mknowlton89) [inline packages/back-end/src/util/scheduleRules.ts:12]: Any insight on a better way to handle this? Typescript kept throwing an error on the line below since s.timestamp could be a `string | null`. Though, that will never happen since we're filtering nulls out just above this. 

I tried to create a new type, but unfortunately that didn't work.

Comment (bttf) [reply packages/back-end/src/util/scheduleRules.ts:12]: This is one way to force it, I'm sure there are others ... Should be ok as long as we are sure it's a string

```
    .map((s) => ({
      date: new Date(s.timestamp as string),
```

Comment (mknowlton89) [reply packages/back-end/src/util/scheduleRules.ts]: Thank you!

**Review by bttf — DISMISSED** [review]

File: packages/back-end/types/feature.d.ts [hunk 34]
```diff
@@ -49,7 +49,7 @@ export interface FeatureInterface {
 }
 type ScheduleRule = {
   timestamp: string | null;
-  enableFeature: boolean;
+  status: "enabled" | "disabled";
```
Comment (jdorn) [inline packages/back-end/types/feature.d.ts:52]: Are there ever going to be more than 2 statuses here?  If not, this seems like it should be a boolean

File: packages/back-end/src/util/scheduleRules.ts [hunk 35]
```diff
@@ -1,46 +1,33 @@
...
-      dateB = b.enableFeature ? new Date("01-01-1971") : new Date("01-01-2090"); //TODO: Find a better way to handle this
-    } else {
-      dateB = new Date(b.timestamp);
-    }
+  const sorted = scheduledRules
+    // Remove nulls
+    .filter((s) => s.timestamp)
+    // Convert timestamps to date objects
+    .map((s) => ({
+      date: new Date(s.timestamp as string),
```
Comment (jdorn) [inline packages/back-end/src/util/scheduleRules.ts:12]: This is fine, but to answer your question about how to make typescript understand that you're filtering out nulls, you can use type guards (also called type predicates) - https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates

```ts
// Type after the filter is applied
type ScheduleRuleNoNulls = Omit<ScheduleRule, "timestamp"> & { timestamp: Date }

const sorted = scheduledRules
  // Remove nulls
  .filter((s): s is ScheduleRuleNoNulls => !!s.timestamp)
  // `s.timestamp` is now Date, not Date|null
  .map((s) => ({
    date: new Date(s.timestamp)
```

Using `as` instead like you have it works, but is basically telling Typescript to ignore the type error that it sees.  If you later remove that `filter`, using `as` will continue to compile just fine, but will throw a runtime error.  Using type predicates, if you remove that filter it will throw a type error at build time instead, which is easier to catch.

Again, not a big deal in this instance since it's very unlikely to break in the future.  This is just for future reference.

File: packages/back-end/types/feature.d.ts (hunk 34 above)
Comment (mknowlton89) [reply packages/back-end/types/feature.d.ts]: Updated to 

```
type ScheduleRule = {
   timestamp: string | null;
   scheduleRuleEnabled: boolean;
}

File: packages/front-end/components/Features/ScheduleInputs.tsx [hunk 36]
```diff
@@ -86,93 +87,137 @@ export default function ScheduleInputs(props: Props) {
...
+          <li className={styles.listitem}>
+            <div className="row align-items-center">
+              <span className="ml-2 mb-2">Launch rule</span>
+              <div className="col-sm-12 col-md mb-2 pl-2 pr-2">
+                <SelectField
+                  name="date-operator"
+                  value={
+                    rules[0].timestamp === null
+                      ? "immediately"
+                      : "at a specific date and time"
```
Comment (jdorn) [inline packages/front-end/components/Features/ScheduleInputs.tsx:99]: It's hard to tell by looking at this code if this will work or not.  You have to match a hard-coded string here against one on line 38.  I think it would be easier to read if you define these options inline.

Also, there's no reason to use this user-facing language as the option value.  You could instead do something like this:

```ts
[
  { label: "immediately", value: "" },
  { label: "at a specific date and time", value: "timestamp" },
]
```

That way you wouldn't have the hard-coded string `at a specific date and time` all throughout the code.

```ts
if (value === "at a specific date and time") {
```

becomes:

```ts
if (value) {
```

File: packages/front-end/components/Features/ScheduleInputs.tsx [hunk 37]
```diff
@@ -86,93 +87,137 @@ export default function ScheduleInputs(props: Props) {
...
+                      name="timestamp"
                     />
+                    <span className="pl-2">
+                      (
+                      {new Date()
+                        .toLocaleDateString(undefined, {
+                          day: "2-digit",
+                          timeZoneName: "short",
+                        })
+                        .substring(4)}
```
Comment (jdorn) [inline packages/front-end/components/Features/ScheduleInputs.tsx:140]: This doesn't work for all locales.  For example, in South Korea (`ko-KR`), a "2-digit" day is actually 3 characters long so your substring will be incorrect.  Timezones and locales have so many edge cases I don't recommend touching them at all without a proper date library.

File: packages/back-end/src/util/scheduleRules.ts (hunk 35 above)
Comment (mknowlton89) [reply packages/back-end/src/util/scheduleRules.ts:12]: I went ahead and updated to use the type guard just for familiarity so if I need to do this again in the future I have a better concept.

File: packages/front-end/components/Features/ScheduleInputs.tsx (hunk 36 above)
Comment (mknowlton89) [reply packages/front-end/components/Features/ScheduleInputs.tsx:99]: Great feedback - this does ready a lot easier after applying these changes.

File: packages/front-end/components/Features/ScheduleInputs.tsx (hunk 37 above)
Comment (mknowlton89) [reply packages/front-end/components/Features/ScheduleInputs.tsx:140]: I'm doing something similar within the `RuleStatusPill` component - I'll update that there, as well.

Comment (mknowlton89) [reply packages/front-end/components/Features/ScheduleInputs.tsx:140]: Updated to use the `date-fns-tz` library.

**Review by jdorn — APPROVED** [review]
